pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openmrs.analytics;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.nio.file.DirectoryNotEmptyException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.io.fs.MatchResult.Status;
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.support.CronExpression;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class is responsible for purging the DWH snapshots when they expire. It also exposes few
+ * helper methods to access the DWH files and directories.
+ */
+@EnableScheduling
+@Component
+public class DWHFilesManager {
+
+  @Autowired private DataProperties dataProperties;
+
+  private CronExpression purgeCron;
+
+  private LocalDateTime lastPurgeRunEnd;
+
+  private volatile boolean isPurgeJobRunning;
```

Comment:
Does this save us a lot? Why not avoid `volatile` and use `synchronized` like everywhere else (I mean even in this class you still need to use `synchronized`)?

Reply:
We could make the `purgeDwhFiles` method synchronized, which means any two threads trying to access this method for the same object will get serialised (one will wait for the other to get completed). Since this is a periodic job and the throughput is less, synchronisation should be fine. However, just to make to little optimal I introduced this lock behaviour so that whichever thread gets the lock will execute the job and the other thread will simply end (no waiting).

Reply:
I see, it is fine if you don't want to change `purgeDwhFiles` to `synchronized` but access to `isPurgeJobRunning` is `synchronized` already, right? I mean, just dropping `volatile` here does not hurt any concurrent logic, does it?

Reply:
You are right, dropping the volatile keyword does not impact anything. Have done the changes.